src/templates.rs

```diff
@@ -275,6 +276,83 @@ pub fn register_case_helpers<'a>(mut h: Handlebars<'a>) -> Handlebars<'a> {
     h
 }
 
+#[derive(Clone, Copy)]
+pub struct FilterHelper;
+
+/// A Handlebars helper to filter an iterable JSON value.
+/// It receives the value to be filtered and a string containing the condition predicate,
+/// then uses Handlebars' truthy logic to filter the items in the value.
+/// It also supports the `#if` helper's `includeZero` optional parameter.
+impl HelperDef for FilterHelper {
+    fn call_inner<'reg: 'rc, 'rc>(
+        &self,
+        h: &Helper<'reg, 'rc>,
+        r: &'reg Handlebars<'reg>,
+        _ctx: &'rc Context,
+        _rc: &mut RenderContext<'reg, 'rc>,
+    ) -> Result<ScopedJson<'reg, 'rc>, RenderError> {
+        let value_param = h
+            .param(0)
+            .ok_or(RenderError::new("Filter helper: Param not found for index 0; must be value to be filtered"))?;
+
+        let value = value_param.value();
+
+        let condition_param = h
+            .param(1)
+            .ok_or(RenderError::new("Filter helper: Param not found for index 1; must be string containing filter condition predicate"))?;
+
+        let condition = condition_param
+            .value()
+            .as_str()
+            .ok_or(RenderError::new("Filter helper: filter condition predicate must be a string"))?;
+
+        let include_zero = h
+            .hash_get("includeZero")
+            .and_then(|v| v.value().as_bool())
+            .unwrap_or(false);
+
+        let items: Vec<Value> = match value {
+            Value::Array(items) => Ok(
+                items
+                    .iter()
+                    .map(|item| item.clone())
+                    .collect()
+            ),
+            // FIXME: This doesn't preserve object keys.
+            // That's probably unexpected for consumers.
```

Comment:
To fix this you could perhaps consider creating a `FilterableValue` enum with the following variants:

```rs
pub enum FilterableValues {
    Array(Vec<Value>),
    Object(Map<String, Value>)
}
```

Then instead of assigning items as a `Vec<Values>` you replace that with the new enum:

```rs
let items: FilterableValues = match value {
    Value::Array(items) => Ok(FilterableValues::Array(items.iter().cloned().collect())),
    Value::Object(items) => Ok(FilterableValues::Object(items.clone())),
     _ => Err(
         RenderError::new("Filter helper: value to be filtered must be an array or object")
     ),
  }?;
```

then each variant can be handled like so:

```rs
match items {
    FilterableValues::Array(items) => {
        let mut filtered_array = vec![];
        for item in items.iter() {
            match r.render_template(&template, item) {
                Ok(s) => {
                    if s.as_str() == "true" {
                        filtered_array.push(item)
                    }
                }
                Err(e) => {
                    return Err(e);
                }
            }
        }
        Ok(ScopedJson::Derived(json!(filtered_array)))
    },
    FilterableValues::Object(object) => {
        let mut filtered_object = Map::new();
        for key in object.keys() {
            if let Some(v) = object.get(key) {
                if v.as_str().ok_or(RenderError::new("something"))? == "true" {
                    filtered_object.insert(key.into(), v.clone());
                }
            }
        }
        Ok(ScopedJson::Derived(json!(filtered_object)))
    }
}
```

Note that I have not really tested this code, so take this more as a recommendation, tell me what you think

Reply:
This is both a review and an education; thank you for all this!

Reply:
I've implemented the suggestions and it compiles; I haven't tested it yet though (still trying to remember how I tested it in the past!)

Reply:
You can test it out in with by compiling a handlebars template as a string, similar to how it was done here

https://github.com/holochain/scaffolding/blob/develop-0.2/src/templates/helpers/merge.rs#L279